http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/AwsCrtAsyncHttpClient.java

```diff
@@ -343,6 +359,34 @@ Builder connectionHealthChecksConfiguration(Consumer<ConnectionHealthChecksConfi
          * Configure the maximum amount of time that a connection should be allowed to remain open while idle.
          */
         Builder connectionMaxIdleTime(Duration connectionMaxIdleTime);
+
+        /**
+         * Configure socket options configuration for all connections established by this client.
+         *
+         * <p>
+         * eg: you can override socket domains and types,  connection timeouts,
+         * and enable periodic keepalive packets which are disabled out of the box, including keepalive timeouts
+         * This may be required for certain connections for longer durations than default socket timeouts
+         *
+         * @param socketOptionsConfiguration The socket configuration to use
+         * @return The builder of the method chaining.
+         */
+        Builder socketOptionsConfiguration(SocketOptionsConfiguration socketOptionsConfiguration);
```

Comment:
This does not seem to be consistent with the APIs we have in other HTTP clients(netty and apache). I wonder if we should add more common settings like [connectionTimeout](https://github.com/aws/aws-sdk-java-v2/blob/master/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClient.java#L280), [tcpKeepAlive](https://github.com/aws/aws-sdk-java-v2/blob/master/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClient.java#L383) in the top level and then introduce
something like putSocketOption(SocketOption socketOption, Object value) for more "advanced setting" just like how we did in netty. This way, we don't need to make code changes if there is a new socket option and keep the number of settings less complex.


https://github.com/aws/aws-sdk-java-v2/blob/d6b02d51a69842957f19f87d8f6ef3797ad791fa/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClient.java#L396

Reply:
If we remove `SocketDomain` and `SocketType`, I think we can add `tcpKeepAlive` and `connectTimeout` to the top level builder

Reply:
I like consistency. But is there a top-level builder that is shared between Netty and Crt already?  What about keepAliveInterval? And are SocketDomain and SocketType worth exposing for configuration for other use-case (I don't have a case today)

In either case it definitely makes sense to make keepAlive a boolean since having "0" mean "off" is not idiomatic for a Java SDK